### PR TITLE
Update `prover-client` to use `ProofKey` instead of `Uuid`

### PR DESCRIPTION
**Pull request by prajwolrg**
## Description

Currently, the prover-client uses Uuid to track proving tasks, which makes it difficult to understand the nature of the task at a glance. The objective of this ticket is to replace the Uuid-based tracking system with a `ProofContext`. 

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature/Enhancement (non-breaking change which adds functionality or enhances an existing one)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Documentation update
- [X] Refactor
- [X] New or updated tests
- [X] Dependency Update

## Checklist

- [ ] I have performed a self-review of my code.
- [ ] I have commented my code where necessary.
- [ ] I have updated the documentation if needed.
- [ ] My changes do not introduce new warnings.
- [ ] I have added tests that prove my changes are effective or that my feature works.
- [ ] New and existing tests pass with my changes.

### CODE REVIEW

**Review by evgenyzdanovich — DISMISSED** [review]
Nice work, LGTM!

File: bin/prover-client/src/handlers/cl_stf.rs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        Ok(cl_stf_id_buf.into())
+    }
+
+    pub async fn get_slot(&self, id: L2BlockId) -> Result<u64, ProvingTaskError> {
+        let header: RpcBlockHeader = self
+            .cl_client
+            .request("strata_getHeaderById", rpc_params![id])
+            .await
+            .map_err(|e| ProvingTaskError::RpcError(e.to_string()))?;
+        Ok(header.block_idx)
```
Comment (delbonis) [inline bin/prover-client/src/handlers/cl_stf.rs:58]: Should be using the RPC API via `jsonrpsee` instead of manually implementing a JSON-RPC client again.

File: bin/prover-client/src/handlers/cl_stf.rs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        let block_id = match task_id.context() {
+            ProofContext::ClStf(id) => id,
+            _ => return Err(ProvingTaskError::InvalidInput("EvmEe".to_string())),
+        };
+        let block_num = self.get_slot(*block_id).await?;
+        let raw_witness: Option<Vec<u8>> = self
+            .cl_client
+            .request("strata_getCLBlockWitness", rpc_params![block_num])
+            .await
+            .map_err(|e| ProvingTaskError::RpcError(e.to_string()))?;
```
Comment (delbonis) [inline bin/prover-client/src/handlers/cl_stf.rs:104]: See other comment.

File: bin/prover-client/src/handlers/checkpoint.rs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            l1_batch_dispatcher,
+            l2_batch_dispatcher,
+        }
+    }
+
+    async fn fetch_info(&self, ckp_idx: u64) -> Result<RpcCheckpointInfo, ProvingTaskError> {
+        self.cl_client
+            .request::<Option<RpcCheckpointInfo>, _>(
+                "strata_getCheckpointInfo",
+                rpc_params![ckp_idx],
```
Comment (delbonis) [inline bin/prover-client/src/handlers/checkpoint.rs:44]: See other comment.

**Review by delbonis — CHANGES_REQUESTED** [review]
The term "handler" might be incorrect for all the types in the `handlers` mod, since it implies some inversion of control where you construct the handler and pass it to a runtime that invokes it like a callback.  It seems more like you could call them "providers" or "repository interfaces" something?

**Comment by prajwolrg** [comment]
> The term "handler" might be incorrect for all the types in the `handlers` mod, since it implies some inversion of control where you construct the handler and pass it to a runtime that invokes it like a callback. It seems more like you could call them "providers" or "repository interfaces" something?

Renamed all handlers -> operators.

**Comment by delbonis** [comment]
Why "operators"?

**Comment by prajwolrg** [comment]
> Why "operators"? I don't really understand the reasoning.

The trait is defined as `ProvingOperation` which is meant to define the operations required for proof generation.
Since this trait is being implemented renaming it to Operators.

File: functional-tests/run_test.sh [hunk 4]
```diff
@@ -11,9 +11,8 @@ fi
 
 # Conditionally run cargo build based on PROVER_TEST
 if [ ! -z $PROVER_TEST ]; then
-    echo "Running on prover mode"
-    # cargo build -F "prover"
-    cargo build -F "prover" --release
+    echo "Running on risc0-mock mode"
+    cargo build --release -F risc0-mock
```
Comment (MdTeach) [inline functional-tests/run_test.sh:15]: With this PR, we can run the current prover client tests in native mode right ?

File: crates/zkvm/adapters/sp1/src/host.rs [hunk 5]
```diff
@@ -66,7 +66,7 @@ impl ZkVmHost for SP1Host {
             std::env::set_var("SP1_PROVER", "mock");
         }
 
-        sp1_sdk::utils::setup_logger();
+        // sp1_sdk::utils::setup_logger();
```
Comment (MdTeach) [inline crates/zkvm/adapters/sp1/src/host.rs:69]: NIT: We can’t use this at runtime, correct? Should we remove it, or wrap it inside cfg(test)?

File: functional-tests/run_test.sh (hunk 4 above)
Comment (prajwolrg) [reply functional-tests/run_test.sh:15]: Yes, we can do that. Can we scope that to a different PR as this is already getting large.
When writing new functional tests, we also want to cover edge cases related with bridge deposits and withdraws as well.

File: crates/zkvm/adapters/sp1/src/host.rs (hunk 5 above)
Comment (prajwolrg) [reply crates/zkvm/adapters/sp1/src/host.rs]: Yup makes sense to drop it.

File: bin/prover-client/src/args.rs [hunk 6]
```diff
@@ -41,6 +51,36 @@ pub struct Args {
...
+        description = "number of sp1 prover workers to spawn",
+        default = "20"
+    )]
+    pub sp1_workers: usize,
+
+    #[argh(
+        option,
+        description = "number of native prover workers to spawn",
+        default = "20"
+    )]
```
Comment (storopoli) [inline bin/prover-client/src/args.rs:74]: Damn 20?!
Whats the rationale for that?
This is like 20 threads no?

File: bin/prover-client/src/args.rs [hunk 7]
```diff
@@ -66,4 +106,22 @@ impl Args {
...
+            workers.insert(ProofZkVm::SP1, self.sp1_workers);
+        }
+
+        #[cfg(feature = "risc0")]
+        {
+            workers.insert(ProofZkVm::Risc0, self.risc0_workers);
+        }
+
+        workers
+    }
```
Comment (storopoli) [inline bin/prover-client/src/args.rs:126]: I know we discussed this. I gave some thought. I still don't like implementation details to be hardcoded in abstractions like traits.

This should be moved into the `impl`s themselves:

```suggestion

    pub fn get_workers(&self) -> HashMap<ProofZkVm, usize>;
```

Comment (storopoli) [reply bin/prover-client/src/args.rs:126]: Also docstrings. VERY IMPORTANT in traits.

File: bin/prover-client/src/errors.rs [hunk 8]
```diff
@@ -1,46 +1,46 @@
+use strata_db::DbError;
+use strata_primitives::proof::ProofKey;
+use strata_zkvm::ZkVmError;
 use thiserror::Error;
 
+use crate::status::ProvingTaskStatus;
+
 // Define custom error type
 #[derive(Error, Debug)]
 pub enum ProvingTaskError {
```
Comment (storopoli) [inline bin/prover-client/src/errors.rs:10]: This comment needs to be expanded into a docstring.
Also can you add docstrings to the variants?

Comment (storopoli) [reply bin/prover-client/src/errors.rs:10]: I know that the error messages are really helpful, but they don't show up in the LSP (IDE autocompletions) or documentation.

File: bin/prover-client/src/hosts/mod.rs [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+use strata_zkvm::{VerificationKey, ZkVmHost};
+
+pub mod native;
+#[cfg(feature = "risc0")]
+pub mod risc0;
+#[cfg(feature = "sp1")]
+pub mod sp1;
+
+pub fn get_verification_key(key: &ProofKey) -> VerificationKey {
+    match key.host() {
```
Comment (storopoli) [inline bin/prover-client/src/hosts/mod.rs:12]: docstring

File: bin/prover-client/src/hosts/risc0.rs [hunk 10]
```diff
@@ -26,13 +25,13 @@ static CL_AGG_HOST: LazyLock<Risc0Host> =
...
 
-pub fn get_host(vm: ProofVm) -> &'static Risc0Host {
-    match vm {
-        ProofVm::BtcProving => &BTC_BLOCKSPACE_HOST,
-        ProofVm::L1Batch => &L1_BATCH_HOST,
-        ProofVm::ELProving => &EVM_EE_STF_HOST,
-        ProofVm::CLProving => &CL_STF_HOST,
-        ProofVm::CLAggregation => &CL_AGG_HOST,
-        ProofVm::Checkpoint => &CHECKPOINT_HOST,
+pub fn get_host(id: &ProofContext) -> &'static Risc0Host {
```
Comment (storopoli) [inline bin/prover-client/src/hosts/risc0.rs:28]: docstring

File: bin/prover-client/src/hosts/sp1.rs [hunk 11]
```diff
@@ -53,13 +52,13 @@ pub static CHECKPOINT_HOST: LazyLock<SP1Host> = std::sync::LazyLock::new(|| {
...
 
-pub fn get_host(vm: ProofVm) -> &'static SP1Host {
-    match vm {
-        ProofVm::BtcProving => &BTC_BLOCKSPACE_HOST,
-        ProofVm::L1Batch => &L1_BATCH_HOST,
-        ProofVm::ELProving => &EVM_EE_STF_HOST,
-        ProofVm::CLProving => &CL_STF_HOST,
-        ProofVm::CLAggregation => &CL_AGG_HOST,
-        ProofVm::Checkpoint => &CHECKPOINT_HOST,
+pub fn get_host(id: &ProofContext) -> &'static SP1Host {
```
Comment (storopoli) [inline bin/prover-client/src/hosts/sp1.rs:55]: docstring

File: bin/prover-client/src/operators/btc.rs [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    btc_client: Arc<BitcoinClient>,
+}
+
+impl BtcBlockspaceOperator {
+    /// Creates a new BTC operations instance.
+    pub fn new(btc_client: Arc<BitcoinClient>) -> Self {
+        Self { btc_client }
+    }
+
+    pub async fn get_id(&self, block_num: u64) -> Result<L1BlockId, ProvingTaskError> {
```
Comment (storopoli) [inline bin/prover-client/src/operators/btc.rs:26]: docstring

File: bin/prover-client/src/operators/checkpoint.rs [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+        l2_batch_dispatcher: Arc<ClAggOperator>,
+    ) -> Self {
+        Self {
+            cl_client,
+            l1_batch_dispatcher,
+            l2_batch_dispatcher,
+        }
+    }
+
+    async fn fetch_info(&self, ckp_idx: u64) -> Result<RpcCheckpointInfo, ProvingTaskError> {
```
Comment (storopoli) [inline bin/prover-client/src/operators/checkpoint.rs:41]: docstring

File: bin/prover-client/src/operators/checkpoint.rs [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+        self.cl_client
+            .get_checkpoint_info(ckp_idx)
+            .await
+            .inspect_err(|_| error!(%ckp_idx, "Failed to fetch CheckpointInfo"))
+            .map_err(|e| ProvingTaskError::RpcError(e.to_string()))?
+            .ok_or(ProvingTaskError::WitnessNotFound)
+    }
+
+    pub async fn fetch_latest_ckp_idx(&self) -> Result<u64, ProvingTaskError> {
```
Comment (storopoli) [inline bin/prover-client/src/operators/checkpoint.rs:50]: docstring

File: bin/prover-client/src/operators/cl_stf.rs [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+impl ClStfOperator {
+    /// Creates a new CL operations instance.
+    pub fn new(cl_client: HttpClient, evm_ee_operator: Arc<EvmEeOperator>) -> Self {
+        Self {
+            cl_client,
+            evm_ee_operator,
+        }
+    }
+
+    pub async fn get_id(&self, block_num: u64) -> Result<L2BlockId, ProvingTaskError> {
```
Comment (storopoli) [inline bin/prover-client/src/operators/cl_stf.rs:36]: docstring

File: bin/prover-client/src/operators/cl_stf.rs [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+        let cl_stf_id_buf: Buf32 = l2_headers
+            .expect("invalid height")
+            .first()
+            .expect("at least one l2 blockid")
+            .block_id
+            .into();
+        Ok(cl_stf_id_buf.into())
+    }
+
+    pub async fn get_slot(&self, id: L2BlockId) -> Result<u64, ProvingTaskError> {
```
Comment (storopoli) [inline bin/prover-client/src/operators/cl_stf.rs:53]: docstring

File: bin/prover-client/src/operators/evm_ee.rs [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+    el_client: HttpClient,
+}
+
+impl EvmEeOperator {
+    /// Creates a new EL operations instance.
+    pub fn new(el_client: HttpClient) -> Self {
+        Self { el_client }
+    }
+
+    pub async fn get_block(&self, block_num: u64) -> Result<Block, ProvingTaskError> {
```
Comment (storopoli) [inline bin/prover-client/src/operators/evm_ee.rs:29]: docstring

File: bin/prover-client/src/operators/l1_batch.rs [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+use crate::{errors::ProvingTaskError, hosts, task_tracker::TaskTracker};
+
+#[derive(Debug, Clone)]
+pub struct L1BatchController {
+    btc_client: Arc<BitcoinClient>,
+    btc_blockspace_operator: Arc<BtcBlockspaceOperator>,
+}
+
+impl L1BatchController {
+    pub fn new(
```
Comment (storopoli) [inline bin/prover-client/src/operators/l1_batch.rs:28]: docstring

File: bin/prover-client/src/operators/mod.rs [hunk 19]
```diff
@@ -0,0 +1,150 @@
+//! A module defining traits and operations for proof generation using ZKVMs.
+//!
+//! This module provides the `ProvingOp` trait, which encapsulates the lifecycle of proof generation
+//! tasks. Each proof generation task includes fetching necessary proof dependencies, creating
+//! tasks, fetching inputs, and performing the proof computation using various supported ZKVMs.
+//!
+//! The operations are designed to interact with a `ProofDb` for storing and retrieving proofs,
+//! a `TaskTracker` for managing task dependencies, and `ProofZkVm` hosts for ZKVM-specific
```
Comment (storopoli) [inline bin/prover-client/src/operators/mod.rs:8]:
```suggestion
//! a [`TaskTracker`/ for managing task dependencies, and [`ProofZkVm`] hosts for ZKVM-specific
```
doclink

File: bin/prover-client/src/operators/mod.rs [hunk 20]
```diff
@@ -0,0 +1,150 @@
+//! A module defining traits and operations for proof generation using ZKVMs.
+//!
+//! This module provides the `ProvingOp` trait, which encapsulates the lifecycle of proof generation
```
Comment (storopoli) [inline bin/prover-client/src/operators/mod.rs:3]:
```suggestion
//! This module provides the [`ProvingOp`] trait, which encapsulates the lifecycle of proof generation
```

Nice work with the module-level docs

File: bin/prover-client/src/operators/mod.rs [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+//! tasks. Each proof generation task includes fetching necessary proof dependencies, creating
+//! tasks, fetching inputs, and performing the proof computation using various supported ZKVMs.
+//!
+//! The operations are designed to interact with a `ProofDb` for storing and retrieving proofs,
+//! a `TaskTracker` for managing task dependencies, and `ProofZkVm` hosts for ZKVM-specific
+//! computations.
+//!
+//! Supported ZKVMs:
```
Comment (storopoli) [inline bin/prover-client/src/operators/mod.rs:11]:
```suggestion
//! Supported ZKVMs:

```

empty line around lists

File: bin/prover-client/src/operators/mod.rs [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+
+pub use operator::ProofOperator;
+
+/// A trait defining the operations required for proof generation.
+///
+/// This trait outlines the steps for proof generation tasks, including fetching proof dependencies,
+/// creating tasks, fetching inputs for the prover, and executing the proof computation using
+/// supported ZKVMs.
+pub trait ProvingOp {
+    /// The prover type associated with this operation, implementing the `ZkVmProver` trait.
```
Comment (storopoli) [inline bin/prover-client/src/operators/mod.rs:44]:
```suggestion
    /// The prover type associated with this operation, implementing the [`ZkVmProver`] trait.
```

**Review by storopoli — CHANGES_REQUESTED** [review]

**Review by delbonis — APPROVED** [review]

File: bin/prover-client/src/args.rs (hunk 6 above)
Comment (prajwolrg) [reply bin/prover-client/src/args.rs:74]: Yes. 
We are producing EVM EE & CL STF blocks every ~5s. The proof generation for that was taking roughly 40s in SP1. In 40s we have: 8 EVM EE block + 8 CL STF block -> 16 proofs.
In 40s we also have: 2 BTC block: total 2 proofs.

If we needed to catch up we need to have at least 18 workers.

File: bin/prover-client/src/args.rs (hunk 7 above)
Comment (prajwolrg) [reply bin/prover-client/src/args.rs:126]: I'm a bit confused. This is not a trait.

**Review by MdTeach — DISMISSED** [review]

File: bin/prover-client/src/args.rs (hunk 6 above)
Comment (storopoli) [reply bin/prover-client/src/args.rs]: Can you add a comment on why 20 in the file?

File: bin/prover-client/src/operators/mod.rs [hunk 23]
```diff
@@ -0,0 +1,152 @@
...
+                error!(?e, "Failed to generate proof")
+            }
+        }
+
+        let proof = proof_res.map_err(ProvingTaskError::ZkVmError)?;
+
+        db.put_proof(*task_id, proof)
+            .map_err(ProvingTaskError::DatabaseError)?;
+
+        Ok(())
```
Comment (storopoli) [inline bin/prover-client/src/operators/mod.rs:150]: I know we discussed this. I gave some thought. I still don't like implementation details to be hardcoded in abstractions like traits.

This should be moved into the impls themselves:

```suggestion
    /// Executes the proof computation for the specified task.
    ///
    /// # Arguments
    /// - `task_id`: The key representing the proof task.
    /// - `db`: A reference to the proof database.
    ///
    /// # Returns
    /// An empty result if the proof computation is successful.
    #[instrument(skip(self, db), fields(task_id = ?task_id))]
    async fn prove(&self, task_id: &ProofKey, db: &ProofDb) -> Result<(), ProvingTaskError>;
```

Comment (delbonis) [reply bin/prover-client/src/operators/mod.rs:150]: I agree with this.  This is not what default impls are designed for.  Maybe we should add a lint for that to annoy you and make you rethink it.

Comment (prajwolrg) [reply bin/prover-client/src/operators/mod.rs]: Have tried to solve this. 
PTAL @delbonis @storopoli

**Review by delbonis — APPROVED** [review]

**Review by storopoli — CHANGES_REQUESTED** [review]
Much better now.
Also the `20` rationale is well explained.
Just missing some docstrings and we are good to go on my end.

**Review by storopoli — APPROVED** [review]
ACK dd91e21fab226754506d113a5642d4e176a5c8e6

**Review by MdTeach — APPROVED** [review]

**Review by delbonis — APPROVED** [review]

**Review by AaronFeickert — APPROVED** [review]
Reviewed only for changes relevant to cryptographic primitives. LGTM.
